packages/netsuite-adapter/src/config/bundle_query.ts

```diff
@@ -0,0 +1,109 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import _ from 'lodash'
+import { TypesQuery, FileCabinetQuery, NetsuiteQuery, CustomRecordsQuery } from './query'
+import { BUNDLE_ID_TO_COMPONENTS } from '../autogen/bundle_components/bundle_components'
+import { SuiteAppBundleType } from '../types/bundle_type'
+import { ALL_TYPES_REGEX, INCLUDE_ALL } from './constants'
+import { bundleIdRegex } from '../filters/bundle_ids'
+import { getGroupItemFromRegex } from '../client/utils'
+import { BUNDLE } from '../constants'
+
+export type BundlesQueryAndSupportedBundles = {
+  query: NetsuiteQuery
+  bundlesToInclude: SuiteAppBundleType[]
+}
+
+const getBundleComponents = (bundleId: string, bundleVersion: string | undefined): Set<string> => {
+  if (bundleVersion) {
+    return BUNDLE_ID_TO_COMPONENTS[bundleId][bundleVersion]
+  }
+  // version is unsupported, check all supported versions of this bundle
+  const versionComponentsUnion = new Set(
+    Object.values(BUNDLE_ID_TO_COMPONENTS[bundleId]).flatMap(versionComponents => Array.from(versionComponents)),
+  )
+  return versionComponentsUnion
+}
+
+const buildTypesQuery = (components: Set<string>): TypesQuery => ({
+  isTypeMatch: () => true,
+  areAllObjectsMatch: () => false,
+  isObjectMatch: ({ instanceId }) => !components.has(instanceId),
+})
+
+const buildFileCabinetQuery = (bundlesInfo: SuiteAppBundleType[]): FileCabinetQuery => {
+  const bundleIdSet = new Set(bundlesInfo.map(bundleInfo => bundleInfo.id))
+  return {
+    isFileMatch: filePath => {
+      const bundleId = getGroupItemFromRegex(filePath, bundleIdRegex, BUNDLE)
+      return !(bundleId.length > 0 && bundleIdSet.has(bundleId[0]))
+    },
+    isParentFolderMatch: () => true,
+    areSomeFilesMatch: () => true,
+  }
+}
+
+const buildCustomRecordQuery = (components: Set<string>): CustomRecordsQuery => ({
+  isCustomRecordTypeMatch: () => true,
+  areAllCustomRecordsMatch: () => false,
+  isCustomRecordMatch: ({ instanceId }) => !components.has(instanceId),
+})
+
+const getBundlesToExclude = (
+  installedBundles: SuiteAppBundleType[],
+  bundlesToExclude: string[],
+): [SuiteAppBundleType[], SuiteAppBundleType[]] => {
+  if (bundlesToExclude.length === 0) {
+    return [[], installedBundles]
+  }
+  if (bundlesToExclude.includes(INCLUDE_ALL)) {
+    return [installedBundles, []]
+  }
+  const bundleMatchers = bundlesToExclude.map(
+    matcher => new RegExp(matcher === INCLUDE_ALL ? ALL_TYPES_REGEX : matcher),
```

Comment:
this is redundant because of the condition above:
```suggestion
    matcher => new RegExp(matcher),
```